src/streamlink/plugins/youtube.py

```diff
@@ -214,84 +210,39 @@ def _create_adaptive_streams(self, info, streams, protected):
+            return m.group("video_id")
 
+        res = http.get(url)
         datam = _ytdata_re.search(res.text)
         if datam:
             data = parse_json(datam.group(1))
             # find the videoRenderer object, where there is a LVE NOW badge
+            for vid_ep in search_dict(data, 'currentVideoEndpoint'):
+                video_id = vid_ep.get("watchEndpoint", {}).get("videoId")
+                if video_id:
+                    log.debug("Video ID from currentVideoEndpoint")
+                    return video_id
             for x in search_dict(data, 'videoRenderer'):
                 for bstyle in search_dict(x.get("badges", {}), "style"):
                     if bstyle == "BADGE_STYLE_TYPE_LIVE_NOW":
                         if x.get("videoId"):
-                            self.logger.debug("Found channel video ID via HTML: {0}", x["videoId"])
+                            log.debug("Video ID from videoRenderer (live)")
                             return x["videoId"]
 
-        else:
-            # fall back on API
-            self.logger.debug("No channel data, falling back to API")
-            match = _channelid_re.search(res.text)
-            if not match:
-                return
-
-            channel_id = match.group(1)
-            self.logger.debug("Found channel_id: {0}".format(channel_id))
-            return self._get_channel_video(channel_id)
-
-    def _get_channel_video(self, channel_id):
-        query = {
-            "channelId": channel_id,
-            "type": "video",
-            "eventType": "live",
-            "part": "id",
-            "key": self.get_option("api_key") or API_KEY
-        }
-        res = http.get(API_SEARCH_URL, params=query, raise_for_status=False)
-        if res.status_code == codes.ok:
-            videos = http.json(res, schema=_search_schema)
-
-            for video in videos:
-                video_id = video["id"]["videoId"]
-                self.logger.debug("Found video_id: {0}".format(video_id))
-                return video_id
-        else:
-            try:
-                errors = http.json(res, exception=ValueError)
-                self.logger.error("Cloud not resolve channel video:")
-                for error in errors['error']['errors']:
-                    self.logger.error("  {message} ({reason})".format(**error))
-
-            except ValueError:
-                self.logger.error("Cloud not resolve channel video: {0} error".format(res.status_code))
-
-    def _find_canonical_stream_info(self):
-        res = http.get(self.url)
-        match = _livechannelid_re.search(res.text)
-        if not match:
-            return
-
-        return self._get_stream_info(match.group(1))
-
-    def _get_stream_info(self, url):
-        match = _url_re.match(url)
-        user = match.group("user")
-        live_channel = match.group("liveChannel")
+        for link in itertags(res.text, 'link'):
+            if link.attributes.get("rel") == "canonical":
```

Comment:
I think `canonical` will actually never find a url that is online,
it will only find offline streams or channels without streams.

the online streams are already found before this.

I don't know if you have a test url where you need this, but I did not need it.

Reply:
Yeah, true. The only example I had was embedded videos, but if you take the video ID from the URL it is not required.